File: shared/TextToSpeech.cs
using System.Speech.Synthesis;
using System.ComponentModel;

namespace Opra.Shared;

public class TextToSpeech : INotifyPropertyChanged
{
    public class Voice
    {
        public string Id { get; set; } = string.Empty;
        public string Name { get; set; } = string.Empty;
        public string Language { get; set; } = string.Empty;
        public bool IsDefault { get; set; }
    }

    public class SpeechSettings
    {
        public float Rate { get; set; } = 0.5f;
        public float Pitch { get; set; } = 1.0f;
        public float Volume { get; set; } = 1.0f;
        public string VoiceId { get; set; } = string.Empty;
    }

    public enum SpeechState
    {
        Stopped,
        Speaking,
        Paused
    }

    // Events
    public event PropertyChangedEventHandler? PropertyChanged;
    public event EventHandler? SpeechStarted;
    public event EventHandler? SpeechFinished;
    public event EventHandler? SpeechPaused;
    public event EventHandler? SpeechResumed;
    public event EventHandler<(float Progress, int CurrentWord, int TotalWords)>? ProgressChanged;

    private SpeechSynthesizer? synthesizer;
    private SpeechSettings settings = new();
    private SpeechState state = SpeechState.Stopped;
    private List<string> chunks = new();
    private int currentChunk = 0;
    private bool isChunked = false;
    private float progress = 0.0f;
    private int currentWordIndex = 0;
    private int totalWords = 0;

    public TextToSpeech()
    {
        Initialize();
    }

    ~TextToSpeech()
    {
        Cleanup();
    }

    public SpeechState State
    {
        get => state;
        private set
        {
            if (state != value)
            {
                state = value;
                OnPropertyChanged(nameof(State));
                OnPropertyChanged(nameof(IsSpeaking));
                OnPropertyChanged(nameof(IsPaused));
            }
        }
    }

    public bool IsSpeaking => state == SpeechState.Speaking;
    public bool IsPaused => state == SpeechState.Paused;

    public SpeechSettings Settings
    {
        get => settings;
        set
        {
            settings = value;
            ApplySettings();
            OnPropertyChanged(nameof(Settings));
        }
    }

    public float Progress
    {
        get => progress;
        private set
        {
            if (Math.Abs(progress - value) > 0.01f)
            {
                progress = value;
                OnPropertyChanged(nameof(Progress));
            }
        }
    }

    public int CurrentWordIndex
    {
        get => currentWordIndex;
        private set
        {
            if (currentWordIndex != value)
            {
                currentWordIndex = value;
                OnPropertyChanged(nameof(CurrentWordIndex));
            }
        }
    }

    public int TotalWords
    {
        get => totalWords;
        private set
        {
            if (totalWords != value)
            {
                totalWords = value;
                OnPropertyChanged(nameof(TotalWords));
            }
        }
    }

    public bool Speak(string text)
    {
        if (string.IsNullOrWhiteSpace(text))
            return false;

        Stop(); // Stop any current speech

        var cleanedText = CleanTextForTTS(text);
        if (string.IsNullOrWhiteSpace(cleanedText))
            return false;

        if (synthesizer == null)
            return false;

        try
        {
            synthesizer.SpeakAsync(cleanedText);
            State = SpeechState.Speaking;
            SpeechStarted?.Invoke(this, EventArgs.Empty);
            return true;
        }
        catch
        {
            return false;
        }
    }

    public bool SpeakChunked(List<string> textChunks, int startChunk = 0)
    {
        if (textChunks == null || textChunks.Count == 0 || startChunk >= textChunks.Count)
            return false;

        Stop(); // Stop any current speech

        chunks = textChunks.ToList();
        currentChunk = startChunk;
        isChunked = true;

        return SpeakCurrentChunk();
    }

    public void Pause()
    {
        if (State == SpeechState.Speaking && synthesizer != null)
        {
            synthesizer.Pause();
            State = SpeechState.Paused;
            SpeechPaused?.Invoke(this, EventArgs.Empty);
        }
    }

    public void Resume()
    {
        if (State == SpeechState.Paused && synthesizer != null)
        {
            synthesizer.Resume();
            State = SpeechState.Speaking;
            SpeechResumed?.Invoke(this, EventArgs.Empty);
        }
    }

    public void Stop()
    {
        if (State != SpeechState.Stopped && synthesizer != null)
        {
            synthesizer.SpeakAsyncCancelAll();
        }

        State = SpeechState.Stopped;
        isChunked = false;
        chunks.Clear();
        currentChunk = 0;
        Progress = 0.0f;
        CurrentWordIndex = 0;
        TotalWords = 0;
    }

    public List<Voice> GetAvailableVoices()
    {
        var voices = new List<Voice>();

        if (synthesizer != null)
        {
            foreach (var voice in synthesizer.GetInstalledVoices())
            {
                var voiceInfo = voice.VoiceInfo;
                voices.Add(new Voice
                {
                    Id = voiceInfo.Id,
                    Name = voiceInfo.Name,
                    Language = voiceInfo.Culture.Name,
                    IsDefault = voiceInfo.Name.Contains("Microsoft") && voiceInfo.Name.Contains("Desktop")
                });
            }
        }

        return voices;
    }

    public bool SetVoice(string voiceId)
    {
        if (synthesizer == null)
            return false;

        try
        {
            var voice = synthesizer.GetInstalledVoices().FirstOrDefault(v => v.VoiceInfo.Id == voiceId);
            if (voice != null)
            {
                synthesizer.SelectVoice(voiceId);
                settings.VoiceId = voiceId;
                return true;
            }
        }
        catch
        {
            // Voice not found or error setting voice
        }

        return false;
    }

    public Voice? GetCurrentVoice()
    {
        if (synthesizer?.Voice != null)
        {
            var voiceInfo = synthesizer.Voice;
            return new Voice
            {
                Id = voiceInfo.Id,
                Name = voiceInfo.Name,
                Language = voiceInfo.Culture.Name,
                IsDefault = voiceInfo.Name.Contains("Microsoft") && voiceInfo.Name.Contains("Desktop")
            };
        }

        return null;
    }

    private bool Initialize()
    {
        try
        {
            synthesizer = new SpeechSynthesizer();
            synthesizer.SpeakCompleted += OnSpeakCompleted;
            synthesizer.SpeakProgress += OnSpeakProgress;
            return true;
        }
        catch
        {
            return false;
        }
    }

    private void Cleanup()
    {
        if (synthesizer != null)
        {
            synthesizer.SpeakAsyncCancelAll();
            synthesizer.SpeakCompleted -= OnSpeakCompleted;
            synthesizer.SpeakProgress -= OnSpeakProgress;
            synthesizer.Dispose();
            synthesizer = null;
        }
    }

    private void ApplySettings()
    {
        if (synthesizer == null)
            return;

        try
        {
            // Set rate (convert from 0.0-1.0 to SAPI range -10 to 10)
            int sapiRate = (int)((settings.Rate - 0.5f) * 20.0f);
            sapiRate = Math.Max(-10, Math.Min(10, sapiRate));
            synthesizer.Rate = sapiRate;

            // Set volume (convert from 0.0-1.0 to SAPI range 0-100)
            int sapiVolume = (int)(settings.Volume * 100.0f);
            sapiVolume = Math.Max(0, Math.Min(100, sapiVolume));
            synthesizer.Volume = sapiVolume;

            // Set voice if specified
            if (!string.IsNullOrEmpty(settings.VoiceId))
            {
                SetVoice(settings.VoiceId);
            }
        }
        catch
        {
            // Error applying settings
        }
    }

    private bool SpeakCurrentChunk()
    {
        if (isChunked && currentChunk < chunks.Count)
        {
            return Speak(chunks[currentChunk]);
        }
        return false;
    }

    private void OnSpeakCompleted(object? sender, SpeakCompletedEventArgs e)
    {
        if (isChunked && currentChunk < chunks.Count - 1)
        {
            // Move to next chunk
            currentChunk++;
            _ = Task.Run(() => SpeakCurrentChunk());
        }
        else
        {
            // All chunks finished or single text finished
            State = SpeechState.Stopped;
            isChunked = false;
            chunks.Clear();
            currentChunk = 0;
            SpeechFinished?.Invoke(this, EventArgs.Empty);
        }
    }

    private void OnSpeakProgress(object? sender, SpeakProgressEventArgs e)
    {
        if (e.Text.Length > 0)
        {
            // Calculate progress based on character position
            float progress = (float)e.CharacterPosition / e.Text.Length;
            Progress = Math.Max(0.0f, Math.Min(1.0f, progress));

            // Estimate word count for progress tracking
            var wordsUpToPosition = e.Text.Substring(0, e.CharacterPosition)
                .Split(new[] { ' ', '\t', '\n', '\r' }, StringSplitOptions.RemoveEmptyEntries);
            var allWords = e.Text.Split(new[] { ' ', '\t', '\n', '\r' }, StringSplitOptions.RemoveEmptyEntries);

            CurrentWordIndex = wordsUpToPosition.Length;
            TotalWords = allWords.Length;

            ProgressChanged?.Invoke(this, (Progress, CurrentWordIndex, TotalWords));
        }
    }

    private string CleanTextForTTS(string text)
    {
        if (string.IsNullOrWhiteSpace(text))
            return "No content available for speech synthesis.";

        var cleaned = text;

        // Remove control characters except newlines and tabs
        cleaned = System.Text.RegularExpressions.Regex.Replace(cleaned, @"[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]", "");

        // Remove zero-width characters
        cleaned = System.Text.RegularExpressions.Regex.Replace(cleaned, @"[\u200B-\u200D\u2060\uFEFF]", "");

        // Replace problematic Unicode spaces with regular spaces
        cleaned = System.Text.RegularExpressions.Regex.Replace(cleaned, @"[\u00A0\u2000-\u200F\u2028-\u202F\u205F-\u206F\u3000]", " ");

        // Clean up multiple spaces
        cleaned = System.Text.RegularExpressions.Regex.Replace(cleaned, @"\s{2,}", " ");

        // Trim whitespace
        cleaned = cleaned.Trim();

        return string.IsNullOrWhiteSpace(cleaned) ? "No content available for speech synthesis." : cleaned;
    }

    protected virtual void OnPropertyChanged(string propertyName)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }
}

File: shared/src/TextToSpeech.cpp
#include "TextToSpeech.h"
#include <algorithm>
#include <thread>
#include <chrono>

#ifdef _WIN32
    #include "TextToSpeech_Windows.cpp"
#elif __APPLE__
    #include "TextToSpeech_macOS.cpp"
#endif

namespace Opra {

class TextToSpeech::Impl {
public:
    Impl() = default;
    ~Impl() = default;
    
    bool speak(const std::string& text) {
        if (text.empty()) return false;
        
        // Clean the text
        std::string cleanedText = cleanTextForTTS(text);
        if (cleanedText.empty()) return false;
        
        // Platform-specific implementation
#ifdef _WIN32
        return speakWindows(cleanedText);
#elif __APPLE__
        return speakMacOS(cleanedText);
#else
        return false; // Unsupported platform
#endif
    }
    
    bool speakChunked(const std::vector<std::string>& chunks, int startChunk) {
        if (chunks.empty() || startChunk >= chunks.size()) return false;
        
        // Store chunks for sequential playback
        this->chunks = chunks;
        this->currentChunk = startChunk;
        this->isChunked = true;
        
        // Start with the first chunk
        return speakCurrentChunk();
    }
    
    void pause() {
        if (state == SpeechState::Speaking) {
#ifdef _WIN32
            pauseWindows();
#elif __APPLE__
            pauseMacOS();
#endif
            state = SpeechState::Paused;
            if (speechPausedCallback) {
                speechPausedCallback();
            }
        }
    }
    
    void resume() {
        if (state == SpeechState::Paused) {
#ifdef _WIN32
            resumeWindows();
#elif __APPLE__
            resumeMacOS();
#endif
            state = SpeechState::Speaking;
            if (speechResumedCallback) {
                speechResumedCallback();
            }
        }
    }
    
    void stop() {
        if (state != SpeechState::Stopped) {
#ifdef _WIN32
            stopWindows();
#elif __APPLE__
            stopMacOS();
#endif
            state = SpeechState::Stopped;
            isChunked = false;
            chunks.clear();
            currentChunk = 0;
        }
    }
    
    SpeechState getState() const {
        return state;
    }
    
    bool isSpeaking() const {
        return state == SpeechState::Speaking;
    }
    
    bool isPaused() const {
        return state == SpeechState::Paused;
    }
    
    void setSettings(const SpeechSettings& settings) {
        this->settings = settings;
        
        // Apply settings to platform-specific implementation
#ifdef _WIN32
        setSettingsWindows(settings);
#elif __APPLE__
        setSettingsMacOS(settings);
#endif
    }
    
    SpeechSettings getSettings() const {
        return settings;
    }
    
    std::vector<Voice> getAvailableVoices() const {
#ifdef _WIN32
        return getAvailableVoicesWindows();
#elif __APPLE__
        return getAvailableVoicesMacOS();
#else
        return {};
#endif
    }
    
    bool setVoice(const std::string& voiceId) {
        settings.voiceId = voiceId;
        
#ifdef _WIN32
        return setVoiceWindows(voiceId);
#elif __APPLE__
        return setVoiceMacOS(voiceId);
#else
        return false;
#endif
    }
    
    Voice getCurrentVoice() const {
#ifdef _WIN32
        return getCurrentVoiceWindows();
#elif __APPLE__
        return getCurrentVoiceMacOS();
#else
        return {};
#endif
    }
    
    float getProgress() const {
        return progress;
    }
    
    int getCurrentWordIndex() const {
        return currentWordIndex;
    }
    
    int getTotalWords() const {
        return totalWords;
    }
    
    void setSpeechStartedCallback(SpeechStartedCallback callback) {
        speechStartedCallback = callback;
    }
    
    void setSpeechFinishedCallback(SpeechFinishedCallback callback) {
        speechFinishedCallback = callback;
    }
    
    void setSpeechPausedCallback(SpeechPausedCallback callback) {
        speechPausedCallback = callback;
    }
    
    void setSpeechResumedCallback(SpeechResumedCallback callback) {
        speechResumedCallback = callback;
    }
    
    void setProgressCallback(ProgressCallback callback) {
        progressCallback = callback;
    }
    
    bool initialize() {
#ifdef _WIN32
        return initializeWindows();
#elif __APPLE__
        return initializeMacOS();
#else
        return false;
#endif
    }
    
    void cleanup() {
#ifdef _WIN32
        cleanupWindows();
#elif __APPLE__
        cleanupMacOS();
#endif
    }
    
    // Callbacks for platform-specific implementations
    void onSpeechStarted() {
        state = SpeechState::Speaking;
        if (speechStartedCallback) {
            speechStartedCallback();
        }
    }
    
    void onSpeechFinished() {
        if (isChunked && currentChunk < chunks.size() - 1) {
            // Move to next chunk
            currentChunk++;
            speakCurrentChunk();
        } else {
            // All chunks finished or single text finished
            state = SpeechState::Stopped;
            isChunked = false;
            chunks.clear();
            currentChunk = 0;
            
            if (speechFinishedCallback) {
                speechFinishedCallback();
            }
        }
    }
    
    void onProgress(float progress, int currentWord, int totalWords) {
        this->progress = progress;
        this->currentWordIndex = currentWord;
        this->totalWords = totalWords;
        
        if (progressCallback) {
            progressCallback(progress, currentWord, totalWords);
        }
    }
    
private:
    SpeechState state = SpeechState::Stopped;
    SpeechSettings settings;
    std::vector<std::string> chunks;
    int currentChunk = 0;
    bool isChunked = false;
    float progress = 0.0f;
    int currentWordIndex = 0;
    int totalWords = 0;
    
    // Callbacks
    SpeechStartedCallback speechStartedCallback;
    SpeechFinishedCallback speechFinishedCallback;
    SpeechPausedCallback speechPausedCallback;
    SpeechResumedCallback speechResumedCallback;
    ProgressCallback progressCallback;
    
    bool speakCurrentChunk() {
        if (isChunked && currentChunk < chunks.size()) {
            return speak(chunks[currentChunk]);
        }
        return false;
    }
    
    std::string cleanTextForTTS(const std::string& text) {
        // Basic text cleaning - more sophisticated cleaning can be added
        std::string cleaned = text;
        
        // Remove control characters except newlines and tabs
        cleaned = std::regex_replace(cleaned, std::regex(R"([\x00-\x08\x0B\x0C\x0E-\x1F\x7F])"), "");
        
        // Clean up multiple spaces
        cleaned = std::regex_replace(cleaned, std::regex(R"(\s{2,})"), " ");
        
        // Trim whitespace
        cleaned = trim(cleaned);
        
        return cleaned;
    }
    
    std::string trim(const std::string& str) {
        size_t first = str.find_first_not_of(' ');
        if (first == std::string::npos) return "";
        size_t last = str.find_last_not_of(' ');
        return str.substr(first, (last - first + 1));
    }
    
    // Platform-specific method declarations
#ifdef _WIN32
    bool speakWindows(const std::string& text);
    void pauseWindows();
    void resumeWindows();
    void stopWindows();
    void setSettingsWindows(const SpeechSettings& settings);
    std::vector<Voice> getAvailableVoicesWindows() const;
    bool setVoiceWindows(const std::string& voiceId);
    Voice getCurrentVoiceWindows() const;
    bool initializeWindows();
    void cleanupWindows();
#elif __APPLE__
    bool speakMacOS(const std::string& text);
    void pauseMacOS();
    void resumeMacOS();
    void stopMacOS();
    void setSettingsMacOS(const SpeechSettings& settings);
    std::vector<Voice> getAvailableVoicesMacOS() const;
    bool setVoiceMacOS(const std::string& voiceId);
    Voice getCurrentVoiceMacOS() const;
    bool initializeMacOS();
    void cleanupMacOS();
#endif
};

// Constructor and destructor
TextToSpeech::TextToSpeech() : pImpl(std::make_unique<Impl>()) {}
TextToSpeech::~TextToSpeech() = default;

// Public method implementations
bool TextToSpeech::speak(const std::string& text) {
    return pImpl->speak(text);
}

bool TextToSpeech::speakChunked(const std::vector<std::string>& chunks, int startChunk) {
    return pImpl->speakChunked(chunks, startChunk);
}

void TextToSpeech::pause() {
    pImpl->pause();
}

void TextToSpeech::resume() {
    pImpl->resume();
}

void TextToSpeech::stop() {
    pImpl->stop();
}

TextToSpeech::SpeechState TextToSpeech::getState() const {
    return pImpl->getState();
}

bool TextToSpeech::isSpeaking() const {
    return pImpl->isSpeaking();
}

bool TextToSpeech::isPaused() const {
    return pImpl->isPaused();
}

void TextToSpeech::setSettings(const SpeechSettings& settings) {
    pImpl->setSettings(settings);
}

TextToSpeech::SpeechSettings TextToSpeech::getSettings() const {
    return pImpl->getSettings();
}

std::vector<TextToSpeech::Voice> TextToSpeech::getAvailableVoices() const {
    return pImpl->getAvailableVoices();
}

bool TextToSpeech::setVoice(const std::string& voiceId) {
    return pImpl->setVoice(voiceId);
}

TextToSpeech::Voice TextToSpeech::getCurrentVoice() const {
    return pImpl->getCurrentVoice();
}

float TextToSpeech::getProgress() const {
    return pImpl->getProgress();
}

int TextToSpeech::getCurrentWordIndex() const {
    return pImpl->getCurrentWordIndex();
}

int TextToSpeech::getTotalWords() const {
    return pImpl->getTotalWords();
}

void TextToSpeech::setSpeechStartedCallback(SpeechStartedCallback callback) {
    pImpl->setSpeechStartedCallback(callback);
}

void TextToSpeech::setSpeechFinishedCallback(SpeechFinishedCallback callback) {
    pImpl->setSpeechFinishedCallback(callback);
}

void TextToSpeech::setSpeechPausedCallback(SpeechPausedCallback callback) {
    pImpl->setSpeechPausedCallback(callback);
}

void TextToSpeech::setSpeechResumedCallback(SpeechResumedCallback callback) {
    pImpl->setSpeechResumedCallback(callback);
}

void TextToSpeech::setProgressCallback(ProgressCallback callback) {
    pImpl->setProgressCallback(callback);
}

bool TextToSpeech::initialize() {
    return pImpl->initialize();
}

void TextToSpeech::cleanup() {
    pImpl->cleanup();
}

} // namespace Opra